http/middleware/log_request_test.go

```diff
@@ -20,47 +24,109 @@ func TestLogRequest(t *testing.T) {
 	require.Equal(t, fmt.Sprintf("%p", middleware.NoopAdapter), fmt.Sprintf("%p", actual))
 
 	// Arrange
+	ip := "192.168.0.0"
+	testID := "test-id"
+	useragent := "trails/test"
+	content := "very-secret/encoding; shhh"
+	referrer := "example.com/referrer"
+	respBody := "test"
+	newExpected := func(expected middleware.LogRequestRecord) middleware.LogRequestRecord {
+		expected.BodySize = len(respBody)
+		expected.Host = "example.com"
+		expected.ID = testID
+		expected.Protocol = "HTTP/1.1"
+		expected.Referrer = referrer
+		expected.ReqContentType = content
+		expected.Status = 200
+		expected.UserAgent = useragent
+
+		return expected
+	}
+
 	tcs := []struct {
 		name     string
 		method   string
 		ip       string
 		url      *url.URL
-		expected string
+		expected middleware.LogRequestRecord
 	}{
-		{"Zero-Value", http.MethodGet, "", &url.URL{Path: "/"}, "GET /"},
-		{"With-IP", http.MethodPost, "127.0.0.1", &url.URL{Path: "/"}, "127.0.0.1 POST /"},
+		{
+			"Zero-Value",
+			http.MethodGet,
+			"",
+			&url.URL{Path: "/"},
+			newExpected(middleware.LogRequestRecord{
+				Method: http.MethodGet,
+				Path:   "/",
+				URI:    "/",
+			}),
+		},
+		{
+			"With-IP",
+			http.MethodPost,
+			ip,
+			&url.URL{Path: "/"},
+			newExpected(middleware.LogRequestRecord{
+				IPAddr: ip,
+				Method: http.MethodPost,
+				Path:   "/",
+				URI:    "/",
+			}),
+		},
 		{
 			"With-Query-Params",
 			http.MethodPut,
-			"0.0.0.0",
+			ip,
 			&url.URL{Path: "/hitting/the/trails", RawQuery: "param=true"},
-			"0.0.0.0 PUT /hitting/the/trails?param=true",
+			newExpected(middleware.LogRequestRecord{
+				IPAddr: ip,
+				Method: http.MethodPut,
+				Path:   "/hitting/the/trails",
+				URI:    "/hitting/the/trails?param=true",
```

Comment:
would be nice to see a test asserting `#fragment` it retained on `uri` as well

Reply:
This sent me down a rabbit trail. Looks like fragments wouldn't be logged in any case? I haven't looked at the source code for how `*http.Request.URL` is formulated, specifically, but I'm seeing fragments are not included in part because the browser doesn't send it (Firefox; but also check out the docstring for [url.ParseRequestURI](https://pkg.go.dev/net/url#ParseRequestURI) which confirms that's the expected behavior).

Accordingly, what I use to construct URI ([url.URL.RequestURI](https://pkg.go.dev/net/url#URL.RequestURI)) doesn't include a fragment even if manually set in a `*url.URL`.

Reply:
Wow, great info. Thanks David